Handle lone surrogates correctly for non-UTF-8 encodings

annevk opened the pull request:
#### 8a954b16d853431cc8dfd069e2994cf4904ca3e3
<pre>
Handle lone surrogates correctly for non-UTF-8 encodings
<a href="https://bugs.webkit.org/show_bug.cgi?id=254888">https://bugs.webkit.org/show_bug.cgi?id=254888</a>
rdar://107530253

Reviewed by Alexey Proskuryakov.

Replace lone surrogates with the replacement code point. Unlike the standard our encoders operate on code points and turning those into scalar values ahead-of-time would be expensive. The next best thing is not surfacing the consequences thereof.

* LayoutTests/fast/url/query-expected.txt:
* LayoutTests/fast/url/query.html:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/urls/resolving-urls/query-encoding/lone-surrogates.sub_encoding=windows-1252-expected.txt:
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):

Canonical link: <a href="https://commits.webkit.org/262513@main">https://commits.webkit.org/262513@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f57cf5d5adeef7b15137db2ae1e7cb08e60ab44a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2603 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1803 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1778 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1775 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1578 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1695 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1522 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1513 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2440 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1517 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1420 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1540 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1541 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2616 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1573 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1401 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1491 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1506 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/430 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1637 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on LayoutTests/fast/url/query-expected.txt:
```diff
@@ -8,7 +8,7 @@ PASS canonicalize('http://www.example.com/?as?df') is 'http://www.example.com/?a
 PASS canonicalize('http://www.example.com/?%40%41123') is 'http://www.example.com/?%40%41123'
 PASS canonicalize('http://www.example.com/?q=你好') is 'http://www.example.com/?q=%26%2320320%3B%26%2322909%3B'
-PASS canonicalize('http://www.example.com/?q=\ud800\ud800') is 'http://www.example.com/?q=%26%2355296%3B%26%2355296%3B'
+PASS canonicalize('http://www.example.com/?q=\ud800\ud800') is 'http://www.example.com/?q=%26%2365533%3B%26%2365533%3B'
```
This result before this PR matches Chrome, although not Firefox.

Is there any effect outside of URL handling, or is this the only scenario where TextCodec::getUnencodableReplacement is called?

annevk replied:
It would impact form submission in a similar way. Except that (should, didn't check in WebKit yet) uses Unencodablehandling::Entities. Surprisingly there's no test coverage for that though. Possibly because it's hard to get lone surrogates in the pipeline to begin with.

annevk replied:
QuestionMarks is dead code: #12304.

annevk replied:
I can't figure out how to make `<form>` hit this with a lone surrogate. And that's the only other scenario there is from a web platform perspective. I did create more test coverage for this in https://github.com/web-platform-tests/wpt/pull/39331.

(I don't understand why TextCodecUTF8::encodeUTF8 takes this enum btw. It doesn't seem to do anything with it: #12311.)

annevk committed:
Handle lone surrogates correctly for non-UTF-8 encodings
https://bugs.webkit.org/show_bug.cgi?id=254888
rdar://107530253

Reviewed by Alexey Proskuryakov.

Replace lone surrogates with the replacement code point. Unlike the standard our encoders operate on code points and turning those into scalar values ahead-of-time would be expensive. The next best thing is not surfacing the consequences thereof.

* LayoutTests/fast/url/query-expected.txt:
* LayoutTests/fast/url/query.html:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/urls/resolving-urls/query-encoding/lone-surrogates.sub_encoding=windows-1252-expected.txt:
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):

Canonical link: https://commits.webkit.org/262513@main
